### PR TITLE
[ONNX] Enable experimental exporter logic to dynamo_export and support refine dynamic_shapes

### LINKED ISSUE #126479: [ONNX] dynamo_export always uses external ONNX weights if the argument is ExportedProgram, even for small models
### 🐛 Describe the bug

dynamo_export always uses external ONNX weights if the argument is ExportedProgram.
dynamo_export never  uses external ONNX weights if the argument is Module - I filed separate issue for that.
This behavior is inconsistent with what onnx.export() had and forces users to calculate memory requirements themselves and branch on that.

### Versions

Pytorch nightly 05/14/24

### LINKED ISSUE #135183: [ONNX] Route dynamo_export with the new logic

### PR DESCRIPTION
**Pull request by titaiwangms**
(1) Enable experimental exporter logic to dynamo_export
(2) Refine dynamic shapes and retry export in export strategies
(3) Delete `torch_export_graph_extractor` and use the new export logic
(4) Disable ExportedProgram test in `test_fx_onnx_with_onnxruntime.py`, as ONNXProgram is different now.

Fixes https://github.com/pytorch/pytorch/issues/126479
Fixes #135183

### CODE REVIEW

**Comment by titaiwangms** [comment]
dynamo_export: delete torch_grap_extracter and when the input is `ExportedProgram`, use the new logic.

File: torch/onnx/__init__.py [hunk 1]
```diff
@@ -396,6 +394,137 @@ def forward(self, x):
...
+        else:
+            dynamic_shapes = None
+
+        return exporter.export_compat(
+            model,  # type: ignore[arg-type]
+            model_args,
+            f=None,
+            kwargs=model_kwargs,
+            dynamic_shapes=dynamic_shapes,
+            opset_version=18,
```
Comment (justinchuby) [inline torch/onnx/__init__.py:515]:
```suggestion
```

Maybe we can leave this as default (I think?)

File: torch/onnx/__init__.py [hunk 2]
```diff
@@ -396,6 +394,137 @@ def forward(self, x):
...
+        if export_options is not None and export_options.dynamic_shapes:
+            # Make all shapes dynamic
+            dynamic_shapes = []
+            for arg_order, model_arg in enumerate(model_args):
+                arg_dynamic_shapes = {}
+                rank = len(model_arg.shape)
+                for idx in range(rank):
+                    dim = torch.export.Dim(f"arg{arg_order}_dim_{idx}")
+                    arg_dynamic_shapes[idx] = dim
+                dynamic_shapes.append(arg_dynamic_shapes)
```
Comment (justinchuby) [inline torch/onnx/__init__.py:505]: model args may be nested I think? There's also keyword args. We need to use Angela's suggestion in https://github.com/pytorch/pytorch/issues/131045 and use pytree

**Review by justinchuby — APPROVED** [review]
Thank you!

File: torch/onnx/_internal/exporter/_capture_strategies.py [hunk 3]
```diff
@@ -120,9 +120,20 @@ class TorchExportStrategy(CaptureStrategy):
...
         self, model, args, kwargs, dynamic_shapes
     ) -> torch.export.ExportedProgram:
-        return torch.export.export(
-            model, args, kwargs=kwargs, dynamic_shapes=dynamic_shapes
-        )
+        try:
+            return torch.export.export(
+                model, args, kwargs=kwargs, dynamic_shapes=dynamic_shapes
+            )
+        except torch._dynamo.exc.UserError as exc:
```
Comment (justinchuby) [inline torch/onnx/_internal/exporter/_capture_strategies.py:127]: Do we need to import `torch._dynamo.exc`?

File: torch/onnx/__init__.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/__init__.py:515]: I can delete this, but it seems the default currently is None

File: torch/onnx/__init__.py (hunk 2 above)
Comment (titaiwangms) [reply torch/onnx/__init__.py:505]: hmmm I didn't figure out a way to get `arg_order` (to compicate the naming), so I didn't choose pytree. We need a way to differentiate dim name though.

File: torch/onnx/_internal/exporter/_capture_strategies.py (hunk 3 above)
Comment (titaiwangms) [reply torch/onnx/_internal/exporter/_capture_strategies.py:127]: It looks like suggested fix only exists under UserError right now.

File: torch/onnx/__init__.py (hunk 1 above)
Comment (justinchuby) [reply torch/onnx/__init__.py:515]: That's ok. We can leave that to the callee

File: torch/onnx/__init__.py [hunk 4]
```diff
@@ -495,14 +496,30 @@ def forward(self, x, bias=None):
...
             # Make all shapes dynamic
-            dynamic_shapes = []
-            for arg_order, model_arg in enumerate(model_args):
-                arg_dynamic_shapes = {}
-                rank = len(model_arg.shape)
-                for idx in range(rank):
-                    dim = torch.export.Dim(f"arg{arg_order}_dim_{idx}")
-                    arg_dynamic_shapes[idx] = dim
-                dynamic_shapes.append(arg_dynamic_shapes)
+            def _to_dynamic_shapes_wrapper():
```
Comment (justinchuby) [inline torch/onnx/__init__.py:499]: We can call it a mapper

File: torch/onnx/_internal/exporter/_compat.py [hunk 5]
```diff
@@ -167,6 +166,20 @@ def export_compat(
             verbose=verbose,
         )
 
+        if f is not None:
```
Comment (titaiwangms) [inline torch/onnx/_internal/exporter/_compat.py:169]: delete this

File: torch/onnx/__init__.py (hunk 4 above)
Comment (titaiwangms) [reply torch/onnx/__init__.py]: Done

**Comment by titaiwangms** [comment]
I will revisit for supporting kwargs to dynamic shapes

**Comment by justinchuby** [comment]
onnx/test_fx_to_onnx.py::TestFxToOnnx::test_save_with_without_initializer_include_initializer_use_fake_mode_use_exported_program Fatal Python error: Segmentation fault

May be how we handle the tensors. May need additional checks when serializing so it doesn't crash
